olp-cpp-sdk-dataservice-read/tests/VersionedLayerClientImplTest.cpp

```diff
@@ -507,4 +548,485 @@ TEST(VersionedLayerClientTest, ProtectThanRelease) {
+          promise.set_value(std::move(response));
+        },
+        nullptr);
+    ASSERT_NE(future.wait_for(std::chrono::seconds(kTimeout)),
+              std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_FALSE(response.IsSuccessful())
+        << response.GetError().GetMessage().c_str();
+    EXPECT_EQ(response.GetError().GetErrorCode(),
+              olp::client::ErrorCode::BadRequest);
+  }
+  {
+    SCOPED_TRACE("Invalid json");
+
+    EXPECT_CALL(*network_mock, Send(IsGetRequest(partitions_path), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(olp::http::NetworkResponse().WithStatus(
+                                         olp::http::HttpStatusCode::OK),
+                                     "invalid json"));
+
+    std::promise<olp::dataservice::read::PrefetchPartitionsResponse> promise;
+    auto future = promise.get_future();
+    auto token = client.PrefetchPartitions(
+        request,
+        [&promise](
+            olp::dataservice::read::PrefetchPartitionsResponse response) {
+          promise.set_value(std::move(response));
+        },
+        nullptr);
+    ASSERT_NE(future.wait_for(std::chrono::seconds(kTimeout)),
+              std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_FALSE(response.IsSuccessful())
+        << response.GetError().GetMessage().c_str();
+    EXPECT_EQ(response.GetError().GetErrorCode(),
+              olp::client::ErrorCode::Unknown);
+    ASSERT_EQ("Fail parsing response.", response.GetError().GetMessage());
+  }
+  {
+    SCOPED_TRACE("Empty data handles");
+
+    auto partitions_response =
+        mockserver::ReadDefaultResponses::GeneratePartitionsResponse(
+            partitions_count);
+    auto& mutable_partitions = partitions_response.GetMutablePartitions();
+    // force empty data handles
+    for (auto& partition : mutable_partitions) {
+      partition.SetDataHandle("");
+    }
+
+    EXPECT_CALL(*network_mock, Send(IsGetRequest(partitions_path), _, _, _, _))
+        .WillOnce(ReturnHttpResponse(
+            olp::http::NetworkResponse().WithStatus(
+                olp::http::HttpStatusCode::OK),
+            olp::serializer::serialize(partitions_response)));
+
+    std::promise<olp::dataservice::read::PrefetchPartitionsResponse> promise;
+    auto future = promise.get_future();
+    auto token = client.PrefetchPartitions(
+        request,
+        [&promise](
+            olp::dataservice::read::PrefetchPartitionsResponse response) {
+          promise.set_value(std::move(response));
+        },
+        nullptr);
+    ASSERT_NE(future.wait_for(std::chrono::seconds(kTimeout)),
+              std::future_status::timeout);
+
+    auto response = future.get();
+    ASSERT_TRUE(response.IsSuccessful())
+        << response.GetError().GetMessage().c_str();
+    const auto result = response.MoveResult();
+    ASSERT_EQ(result.GetPartitions().size(), 0u);
+  }
+}
+
```

Comment:
Add cancel, prefetch progress and request priority tests, please.

Reply:
I think this should be integration tests.